usr/src/uts/common/io/scsi/adapters/smartpqi/smartpqi_init.c

```diff
@@ -0,0 +1,1948 @@
+	s->s_max_oq_element_length = cap->max_oq_element_length * 16;
+
+	iu_layer = &cap->iu_layer_descriptors[PQI_PROTOCOL_SOP];
+	s->s_max_inbound_iu_length_per_firmware =
+	    iu_layer->max_inbound_iu_length;
+	s->s_inbound_spanning_supported = iu_layer->inbound_spanning_supported;
+	s->s_outbound_spanning_supported =
+	    iu_layer->outbound_spanning_supported;
+
+	pqi_free_single(s, dma);
+	return (rval);
+}
+
+static boolean_t
+pqi_valid_device_capability(pqi_state_t s)
+{
+	if (s->s_max_iq_element_length < PQI_OPERATIONAL_IQ_ELEMENT_LENGTH)
+		return (B_FALSE);
+	if (s->s_max_oq_element_length < PQI_OPERATIONAL_OQ_ELEMENT_LENGTH)
+		return (B_FALSE);
+	if (s->s_max_inbound_iu_length_per_firmware <
+	    PQI_OPERATIONAL_IQ_ELEMENT_LENGTH)
+		return (B_FALSE);
+	/* ---- Controller doesn't support spanning but we need it ---- */
+	if (!s->s_inbound_spanning_supported)
+		return (B_FALSE);
+	/* ---- Controller wants outbound spanning, the driver doesn't ---- */
+	if (s->s_outbound_spanning_supported)
+		return (B_FALSE);
+
+	return (B_TRUE);
+}
+
+static boolean_t
+pqi_calculate_queue_resources(pqi_state_t s)
+{
+	int	max_queue_groups;
+	int	num_queue_groups;
+	int	num_elements_per_iq;
+	int	num_elements_per_oq;
+
+	if (reset_devices) {
+		num_queue_groups = 1;
+	} else {
+		max_queue_groups = min(s->s_max_inbound_queues / 2,
+		    s->s_max_outbound_queues - 1);
+		max_queue_groups = min(max_queue_groups, PQI_MAX_QUEUE_GROUPS);
+
+		num_queue_groups = min(ncpus, s->s_intr_cnt);
+		num_queue_groups = min(num_queue_groups, max_queue_groups);
+	}
+	s->s_num_queue_groups = num_queue_groups;
+
+	s->s_max_inbound_iu_length =
+	    (s->s_max_inbound_iu_length_per_firmware /
+	    PQI_OPERATIONAL_IQ_ELEMENT_LENGTH) *
+	    PQI_OPERATIONAL_IQ_ELEMENT_LENGTH;
+
+	num_elements_per_iq = s->s_max_inbound_iu_length /
+	    PQI_OPERATIONAL_IQ_ELEMENT_LENGTH;
+	/* ---- add one because one element in each queue is unusable ---- */
+	num_elements_per_iq++;
+
+	num_elements_per_iq = min(num_elements_per_iq,
+	    s->s_max_elements_per_iq);
+
+	num_elements_per_oq = ((num_elements_per_iq - 1) * 2) + 1;
+	num_elements_per_oq = min(num_elements_per_oq,
+	    s->s_max_elements_per_oq);
+
+	s->s_num_elements_per_iq = num_elements_per_iq;
+	s->s_num_elements_per_oq = num_elements_per_oq;
+
+	s->s_max_sg_per_iu = ((s->s_max_inbound_iu_length -
+	    PQI_OPERATIONAL_IQ_ELEMENT_LENGTH) /
+	    sizeof (struct pqi_sg_entry)) +
+	    PQI_MAX_EMBEDDED_SG_DESCRIPTORS;
+	return (B_TRUE);
+}
+
+static boolean_t
+pqi_alloc_io_resource(pqi_state_t s)
+{
+	pqi_io_request_t	*io;
+	size_t			sg_chain_len;
+	int			i;
+
+	s->s_io_rqst_pool = PQI_ZALLOC(s->s_max_io_slots * sizeof (*io),
+	    KM_SLEEP);
+
+	sg_chain_len = s->s_sg_chain_buf_length;
+	io = s->s_io_rqst_pool;
+	for (i = 0; i < s->s_max_io_slots; i++) {
+		io->io_iu = PQI_ZALLOC(s->s_max_inbound_iu_length, KM_SLEEP);
+
+		/*
+		 * TODO: Don't allocate dma space here. Move this to
```

Comment:
Another `XXX` comment.  If what it says is important, do it before you integrate.  If it's not, remove the comment and file an issue in the bug tracker.